stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4164,6 +4164,71 @@ LogicalResult RngOp::reifyReturnTypeShapes(
 // SelectOp
 //===----------------------------------------------------------------------===//
 
+namespace {
+// Utility function, used by printSelectOpType and
+// printSameOperandsAndResultType
+ParseResult assignFromFunctionType(OpAsmParser& parser, llvm::SMLoc loc,
+                                   ArrayRef<Type*> operands, Type& result,
+                                   FunctionType& fnType) {
+  assert(fnType);
+  if (fnType.getInputs().size() != operands.size()) {
+    return parser.emitError(loc)
+           << operands.size() << " operands present, but expected "
+           << fnType.getInputs().size();
+  }
+  if (fnType.getResults().size() != 1) {
+    return parser.emitError(loc, "expected single output");
+  }
+
+  // Set operand types to function input types
+  for (auto [operand, input] : llvm::zip(operands, fnType.getInputs())) {
+    *operand = input;
+  }
+  result = fnType.getResults().front();
+  return success();
+}
+}  // namespace
+
+void printSelectOpType(OpAsmPrinter& p, Operation* op, Type pred, Type onTrue,
+                       Type onFalse, Type result) {
+  // Print functional type if true/false branches don't match return type.
+  if (onTrue != result || onFalse != result) {
+    p.printFunctionalType(op);
+    return;
+  }
+
+  // Print pred type and result type
+  p << pred << ", " << result;
+}
+
+ParseResult parseSelectOpType(OpAsmParser& parser, Type& pred, Type& onTrue,
+                              Type& onFalse, Type& result) {
+  // Operand and result types are the same, use copy constructor
+  llvm::SMLoc loc = parser.getCurrentLocation();
+  SmallVector<Type> types;
+  if (parser.parseTypeList(types)) {
+    return failure();
+  }
+
+  // Error handling for invalid types
+  // Fail if not two types, or single functional type
+  if (types.size() != 2 &&
+      (types.size() != 1 || !types.front().dyn_cast<FunctionType>())) {
+    return parser.emitError(loc,
+                            "expected functional type or list of two types");
+  }
+
+  if (types.size() == 2) {
+    pred = types.front();
+    onTrue = onFalse = result = types.back();
+    return success();
+  }
+
+  FunctionType fnType = types.front().dyn_cast<FunctionType>();
```

Comment:
It is idiomatic to say `auto` when the type is obvious from the context: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable.

Reply:
Makes sense!